exec/java-exec/src/main/java/org/apache/drill/exec/planner/fragment/LocalAffinityFragmentParallelizer.java

```diff
@@ -0,0 +1,149 @@
+
+        // 6: Finally make sure the width is at least one
+        width = Math.max(1, width);
+
+        List<DrillbitEndpoint> endpointPool = Lists.newArrayList();
+        List<DrillbitEndpoint> assignedEndPoints = Lists.newArrayList();
+
+        Map<DrillbitEndpoint, EndpointAffinity> endpointAffinityMap =
+            fragmentWrapper.getStats().getParallelizationInfo().getEndpointAffinityMap();
+
+        int totalAssigned = 0;
+        int totalWorkUnits = 0;
+
+        // Get the total number of work units and list of endPoints to schedule fragments on
+        for (Map.Entry<DrillbitEndpoint, EndpointAffinity> epAff : endpointAffinityMap.entrySet()) {
+            if (epAff.getValue().getNumLocalWorkUnits() > 0) {
+                totalWorkUnits += epAff.getValue().getNumLocalWorkUnits();
+                endpointPool.add(epAff.getKey());
+            }
+        }
+
+        // Keep track of number of fragments allocated to each endpoint.
+        Map<DrillbitEndpoint, Integer> endpointAssignments = new HashMap<>();
+
+        // Keep track of how many more to assign to each endpoint.
+        Map<DrillbitEndpoint, Integer> remainingEndpointAssignments = new HashMap<>();
+
+        // Calculate the target allocation for each endPoint based on work it has to do
+        // Assign one fragment (minimum) to all the endPoints in the pool.
+        for (DrillbitEndpoint ep : endpointPool) {
+            int targetAllocation = (int) Math.ceil(endpointAffinityMap.get(ep).getNumLocalWorkUnits() * width / parallelizationInfo.getMaxWidth());
+            assignedEndPoints.add(ep);
+            totalAssigned++;
+            remainingEndpointAssignments.put(ep, targetAllocation-1);
+            endpointAssignments.put(ep, 1);
+        }
+
+        // Keep allocating from endpoints in a round robin fashion upto
+        // max(targetAllocation, maxwidthPerNode) for each endpoint and
+        // upto width for all together.
+        while(totalAssigned < width) {
+            int assignedThisRound = 0;
+            for (DrillbitEndpoint ep : endpointPool) {
+                if (remainingEndpointAssignments.get(ep) > 0 &&
+                    remainingEndpointAssignments.get(ep) < parameters.getMaxWidthPerNode()) {
+                    assignedEndPoints.add(ep);
+                    remainingEndpointAssignments.put(ep, remainingEndpointAssignments.get(ep) - 1);
+                    totalAssigned++;
+                    assignedThisRound++;
+                    endpointAssignments.put(ep, endpointAssignments.get(ep) + 1);
+                }
+                if (totalAssigned == width) {
+                    break;
+                }
+            }
+            if (assignedThisRound == 0) {
+                break;
+            }
+        }
+
+        // This is for the case where drillbits are not running on endPoints which have data.
+        // Allocate them from the active endpoint pool.
+        int totalUnAssigned =
```

Comment:
I got all unit and regression tests pass with localAffinity=true.  If this algorithm fails, that is not possible. Also, we are doing this only for the case when drillbits are not running on the nodes which have data.